### DISCUSSION TITLE
Rate-limiting plugin by credential

### DISCUSSION THREAD

**Discussion by Jim-Bar** [Kong/kong#10608] [Help]
Hi,

I am having trouble configuring the rate-limiting plugin to limit by credential (i.e. end user). In my understanding, with a rate-limiting configuration by credential of 5 / sec, if two users using the same consumer (i.e. client ID & client secret) but different authentication tokens perform 5 requests each, then Kong should not limit them.

However in my tests they are limited: firing 6 requests in less than a second with 3 of them using the first token and the other 3 the second token, one request will result in a 429 HTTP code. If however I increase the rate-limit to 6, all the requests pass.

The [documentation of the rate-limiting plugin](https://docs.konghq.com/hub/kong-inc/rate-limiting) says:
> If the entity value for aggregating the limits cannot be determined, the system falls back to `ip`.

But I am not sure how the "*entity value for aggregating the limits*" is determined. I do know that the rate-limiting plugin [relies on some kind of `id`](https://github.com/Kong/kong/blob/9bc3c736c4816838640da25692a2974ea9fb09e6/kong/plugins/rate-limiting/handler.lua#L66) set into the context [by whatever authentication plugin](https://github.com/Kong/kong/pull/5497) is used (in my case OAuth 2 plugin).

How do I ensure this id is set? Is the order in which the plugins are configured important?

Thanks for your help.

**Comment by ghost** [comment]
Can you show me your routing and plugin configurations?

**Reply by Jim-Bar** [reply]
Sure, here is the route configuration:

```
{
    "regex_priority":0,
    "snis":null,
    "headers":null,
    "tags":[
        "admin",
        "audit",
        "auth",
        "boards",
        "users"
    ],
    "created_at":1680593265,
    "paths":[
        "/v1/users/me"
    ],
    "methods":[
        "GET"
    ],
    "sources":null,
    "updated_at":1680593265,
    "strip_path":true,
    "request_buffering":true,
    "response_buffering":true,
    "path_handling":"v0",
    "preserve_host":false,
    "https_redirect_status_code":426,
    "id":"a81c2f60-f067-4245-9d8e-1ac271d8073f",
    "destinations":null,
    "name":"user-v1-me",
    "protocols":[
        "http",
        "https"
    ],
    "hosts":null,
    "service":"user-v1-me"
}
```

The *OAuth 2.0 Authentication* plugin configuration:

```
{
    "consumer":null,
    "config":{
        "reuse_refresh_token":false,
        "auth_header_name":"authorization",
        "scopes":[
            "auditlog:read",
            "board:read",
            "board:write",
            "identity:read"
        ],
        "accept_http_if_already_terminated":true,
        "token_expiration":7200,
        "provision_key":"xOQxyNUs5K7YpKvwgZ2BRY1L6ZpJDHt7",
        "global_credentials":true,
        "persistent_refresh_token":false,
        "enable_password_grant":false,
        "enable_authorization_code":true,
        "enable_implicit_grant":false,
        "pkce":"lax",
        "hide_credentials":false,
        "mandatory_scope":true,
        "anonymous":null,
        "refresh_token_ttl":1209600,
        "enable_client_credentials":false
    },
    "tags":[
        "admin",
        "audit",
        "auth",
        "boards",
        "users"
    ],
    "created_at":1680593265,
    "service":{
        "id":"b4105e16-667f-493b-af97-300fd239ef7e"
    },
    "name":"oauth2",
    "id":"fec573bd-3f57-4452-b97f-a648f24c3b7f",
    "route":null,
    "enabled":true,
    "protocols":[
        "grpc",
        "grpcs",
        "http",
        "https"
    ]
}
```

And the *Rate Limiting* plugin configuration:

```
{
    "consumer":null,
    "config":{
        "policy":"local",
        "second":5,
        "redis_timeout":2000,
        "redis_ssl":false,
        "redis_ssl_verify":false,
        "hour":null,
        "redis_database":0,
        "month":null,
        "year":null,
        "redis_password":null,
        "redis_username":null,
        "hide_client_headers":false,
        "header_name":null,
        "redis_host":null,
        "limit_by":"credential",
        "redis_server_name":null,
        "day":null,
        "minute":null,
        "redis_port":6379,
        "path":null,
        "fault_tolerant":true
    },
    "tags":[
        "admin",
        "audit",
        "auth",
        "boards",
        "users"
    ],
    "created_at":1680593265,
    "service":{
        "id":"b4105e16-667f-493b-af97-300fd239ef7e"
    },
    "route":null,
    "protocols":[
        "grpc",
        "grpcs",
        "http",
        "https"
    ],
    "id":"61e2591b-f05b-4d67-b470-84229fe11cc1",
    "name":"rate-limiting",
    "enabled":true
}
```

**Comment by ghost** [comment]
According to https://konghq.com/blog/engineering/kong-gateway-oauth2#example. I have reproduced your scenario and verified the problem you mentioned.

~~The two users you mentioned in Kong are just the same consumer. Kong does not recognize which user the current request is from, because users are not defined by Kong and they come from an external system that Kong cannot differentiate.~~

If you want Kong to recognize the users of its external systems, I think you should map them as different consumers defined by Kong.

Assuming I have two users, Bob and Clark, in Kong they are the same consumer. When they use their respective tokens to access Kong, the way Kong recognizes their identities is handled as follows. 

https://github.com/Kong/kong/blob/0cbe37a393beb29061977de1127a32d0c03e7fed/kong/plugins/rate-limiting/handler.lua#L65-L67

Bob and Clark have the same `identifier` here (because they are the same consumer). 

But I found that Kong can distinguish different users based on the token used by each user. The key point is here:
https://github.com/Kong/kong/blob/0cbe37a393beb29061977de1127a32d0c03e7fed/kong/plugins/oauth2/access.lua#L938-L940

If Bob accesses Kong using his own token, `token.authenticated_userid` is Bob. If Clark uses his own token, `token.authenticated_userid` is Clark.

So my idea is:

1. When the `limit_by` option of the rate-limiting plugin is set to `credential`, Kong uses `token.authenticated_userid` to identify different users.
2. Update the rate-limiting plugin to support a new option called `authenticated_userid` in the `limit_by` field, in order to differentiate between different users.

Choose to implement either 1 or 2.

**Reply by ghost** [reply]
cc @bungle

**Reply by Jim-Bar** [reply]
> If you want Kong to recognize the users of its external systems, I think you should map them as different consumers defined by Kong.

Can you point to a resource to learn how to do this mapping? If I want to also apply a rate-limit by consumer, wouldn't that prevent me from doing so?

> So my idea is:
> 
>     1. When the `limit_by` option of the rate-limiting plugin is set to `credential`, Kong uses `token.authenticated_userid` to identify different users.
> 
>     2. Update the rate-limiting plugin to support a new option called `authenticated_userid` in the `limit_by` field, in order to differentiate between different users.
> 
> 
> Choose to implement either 1 or 2.

As I understand it there is no way of filtering by end user as of now (although I don't quite understand how that statement relates to your first suggestion of mapping to different consumers), without making modifications to either Kong or the rate-limiting plugin. Is that right?

**Reply by ghost** [reply]
> Can you point to a resource to learn how to do this mapping? If I want to also apply a rate-limit by consumer, wouldn't that prevent me from doing so?

I thought about it and realized that this would make things more complicated, so I'm not planning to give you any suggestions.

> As I understand it there is no way of filtering by end user as of now (although I don't quite understand how that statement relates to your first suggestion of mapping to different consumers), without making modifications to either Kong or the rate-limiting plugin. Is that right?

Yes, this is a feature that cannot be achieved in the current version of Kong.
